internal/pkg/composable/providers/kubernetessecrets/expiration_cache.go

```diff
@@ -0,0 +1,124 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package kubernetessecrets
+
+import (
+	"sync"
+	"time"
+)
+
+// expirationCache is a store that expires items after time.Now - secret.lastAccess > ttl (if ttl > 0) at Get or List.
+// expirationCache works with *cacheEntry, a pointer struct that wraps secret, instead of secret directly because map
+// structure in standard go library never removes the buckets from memory even after removing all the elements from it.
+// However, since *cacheEntry is a pointer it can be garbage collected when no longer referenced by the GC, such as
+// when deleted from the map. More importantly working with a pointer makes the entry in the map bucket, that doesn't
+// get deallocated, to utilise only 8 bytes on a 64-bit system.
+type expirationCache struct {
+	sync.RWMutex
+	// ttl is the time-to-live for items in the cache
+	ttl time.Duration
+	// items is the underlying cache store.
+	items map[string]*cacheEntry
+}
+
+type cacheEntry struct {
+	s          secret
+	lastAccess time.Time
+}
+
+// Get returns the secret associated with the given key from the store if it exists and is not expired. If updateAccess is true
+// and the secret exists, essentially the expiration check is skipped and the lastAccess timestamp is updated to time.Now().
+func (c *expirationCache) Get(key string, updateAccess bool) (secret, bool) {
+	c.RLock()
+	defer c.RUnlock()
+
+	entry, exists := c.items[key]
+	if !exists {
+		return secret{}, false
+	}
+	if updateAccess {
+		entry.lastAccess = time.Now()
+	} else if c.isExpired(entry.lastAccess) {
+		delete(c.items, key)
+		return secret{}, false
+	}
```

Comment:
This is a modification of the map while holding only a Read lock, this is not concurrent-safe.
There's also a possible concurrent access on `lastAccess` at the entry level.

Reply:
Makes sense to me. "premature optimization is the root of all evil" :wink: